pkg/azure/client.go

```diff
@@ -102,8 +111,40 @@ func (az *azClient) SetAuthorizer(authorizer autorest.Authorizer) {
 	az.deploymentsClient.Authorizer = authorizer
 }
 
-func (az *azClient) GetGateway() (n.ApplicationGateway, error) {
-	return az.appGatewaysClient.Get(az.ctx, string(az.resourceGroupName), string(az.appGwName))
+func (az *azClient) GetGateway() (response n.ApplicationGateway, err error) {
+	err = utils.Retry(retryCount, retryPause,
+		func() (utils.Retriable, error) {
+			response, err = az.appGatewaysClient.Get(az.ctx, string(az.resourceGroupName), string(az.appGwName))
+			if err == nil {
+				return utils.Retriable(false), nil
+			}
+
+			// Reasons for 403 errors
+			if response.Response.Response != nil && response.Response.StatusCode == 403 {
+				glog.Error("Possible reasons:" +
+					" AKS Service Principal requires 'Managed Identity Operator' access on Controller Identity;" +
+					" 'identityResourceID' and/or 'identityClientID' are incorrect in the Helm config;" +
+					" AGIC Identity requires 'Contributor' access on Application Gateway and 'Reader' access on Application Gateway's Resource Group;")
+			}
+
+			if response.Response.Response != nil && response.Response.StatusCode == 404 {
+				glog.Error("Got 404 NOT FOUND status code on getting Application Gateway from ARM.")
+				return utils.Retriable(false), ErrAppGatewayNotFound
+			}
+
+			if response.Response.Response != nil && response.Response.StatusCode != 200 {
+				// for example, getting 401. This is not expected as we are getting a token before making the call.
+				glog.Error("Unexpected ARM status code on GET existing App Gateway config: ", response.Response.StatusCode)
+			}
+
+			glog.Errorf("Failed fetching config for App Gateway instance. Will retry in %v. Error: %s", retryPause, err)
+			return utils.Retriable(true), ErrGetArmAuth
+		})
+
+	if err != nil && err != ErrAppGatewayNotFound {
+		glog.Errorf("Tried %d times to authenticate with ARM; Error: %s", retryCount, err)
+	}
+	return
```

Comment:
Do we not need to return values ?

Reply:
This go's syntactic sugar. If a function uses named return parameters, then return can implicitly know what to return.
https://golang.org/doc/effective_go.html#named-results